content/blog/upkeep-testthat-3/index.Rmd

```diff
@@ -0,0 +1,142 @@
+
+### Comparing things `r emo::ji("apple")` `r emo::ji("orange")`
+
+The last big change from edition 2 to edition 3 that I want to mention is what is happening under the hood of `expect_equal()` and `expect_identical()`. Edition 3 uses `waldo::compare()` while edition 2 uses `all.equal()`. For the most part that meant changing the argument name from `tol` to `tolerance`, like in my third commit above. 
+
+I did however run into the situation where a test newly failed. Those are the situations where general advise is hard because it depends so much on the context. In my case, I made use of the `ignore_function_env` and `ignore_formula_env` arguments to `waldo::compare()` to exclude those environments from the comparison. Those are probably really neat to know about if you are upgrading a modelling package but not particularly important otherwise. For dials and censored, that solved most of the cases. In one instance, I ended up tweaking the reference value based on theoretical considerations of the model I was dealing with rather than increasing the tolerance. 
+
+Those instances may be the most work when upgrading to edition 3 but I did not encounter many of them - and when I did it was valuable to know about the differences (well, those which I didn't choose to ignore).
+
+
+## More testing made easier
+
+While I was going over all the test files, I also decided to cover a few other aspects.
+
+### Nested expectations
+
+When [Davis Vaughan](https://github.com/DavisVaughan) moved other tidymodels packages to testthat 3e, I saw him disentangle nested expectations. For example, patterns like 
+
+```r
+expect_warning(expect_equal(one_call, another_call))
+```
+or 
+```r
+expect_equal(expect_warning(one_call), expect_warning(another_call))
+```
+can be re-written as
+```r
+expect_snapshot({
+    object_from_one_call <- one_call()
+    object_from_another_call <- another_call()
+})
+expect_equal(object_from_one_call, object_from_another_call)
+```
+which separates an expectation about the warnings from the expectation about the value, making it easier to see which part(s) fail. Snapshots can also be particularly helpful in situations where you are trying to test for a combination of warnings, messages, and/or errors because they cover them all.
+
+### Self-contained tests
+
+I wanted to make the tests more self-contained so that a test could run with a single call to `test_that()`. Specifically, I didn't want to have to scroll back up to the top of the file to load any necessary package or find the code that creates helper objects.
+
+Namespacing functions helps with the former, i.e. using `dplyr::mutate()` instead of `library(dplyr)` at the top of the file and later `mutate()` inside of the expression for `test_that()`. 
```

Comment:
Is “namespacing functions” what this is called? (Genuine question, not a rhetorical editorial one—I think I usually call it prefixing, which is probably wrong).

Reply:
I'm not 100% sure that is what it's called that but I've heard several people use it. I'll be verbose also in my language, not just the example!